### ISSUE TITLE
AWS::Serverless transform creating resources that fail W3011 (DeletionPolicy/UpdateReplacePolicy rule)

### ISSUE THREAD

**Issue by TLaue** [aws-cloudformation/cfn-python-lint#1265]
*cfn-lint version: (`cfn-lint --version`)*
0.26.01

*Description of issue.*
The latest version of cfn-lint creates a warning in case of missing Deletion and UpdateReplace policy. This complain appears even though both properties are set for an S3 bucket which is the only resource in our stack wich requires these properties according to https://github.com/aws-cloudformation/cfn-python-lint/pull/1232. This warning did not appear prior to version 0.26.1. What is the difference beween I3011 and W3011?

```
Resources:
  xxxBucket:
    Type: AWS::S3::Bucket
    DeletionPolicy: Delete
    UpdateReplacePolicy: Delete
    Properties:...
```

**Comment by PatMyron** [comment]
> What is the difference beween I3011 and W3011?

W3011 is meant to warn about any resource that is protected from deletion through either an [`UpdateReplacePolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatereplacepolicy.html) or a [`DeletionPolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html) but not protected by both

I3011 is meant to inform about any resource missing [`UpdateReplacePolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatereplacepolicy.html) / [`DeletionPolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html) that could potentially result in data loss based on its resource type:

https://github.com/aws-cloudformation/cfn-python-lint/blob/c3d16457059e03eaaf752f403d1983c1c7be3d0a/src/cfnlint/rules/resources/UpdateReplacePolicyDeletionPolicyOnStatefulResourceTypes.py#L22-L39

---

@TLaue are you comfortable sharing the template that is generating the warning? I wasn't able to reproduce with either of these templates:

```yaml
Resources:
  Bucket:
    Type: AWS::S3::Bucket
    DeletionPolicy: Delete
    UpdateReplacePolicy: Delete
```

```yaml
Resources:
  Bucket:
    Type: AWS::S3::Bucket
    DeletionPolicy: Delete
    UpdateReplacePolicy: Delete
    Properties:
      BucketName: bucket
```

**Comment by TLaue** [comment]
May I send the template via email to you? Can you provide an email address?

**Comment by PatMyron** [comment]
@TLaue pamyron@amazon.com

**Comment by TLaue** [comment]
It is on the way.

**Comment by PatMyron** [comment]
@TLaue I believe the `AWS::Serverless` transform is creating [`AWS::Lambda::Version`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-version.html) resources with `DeletionPolicy: Retain` but no [`UpdateReplacePolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatereplacepolicy.html). Will reach out to them about that

**Comment by senior88oqz** [comment]
> AWS::Lambda::Version

In my template, I've got `Transform: AWS::Serverless-2016-10-31` and `AutoPublishAlias: live` property in my `AWS::Serverless::Function`, and it fails to pass the updated `W3011`. I believe it's the `AWS::Lambda::Version` causing the problem, as if I comment out `AutoPublishAlias: live`, `W3011` passes.

**Comment by PatMyron** [comment]
#1266 should make what's happening clearer

@keetonian @praneetap any thoughts on which [`UpdateReplacePolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatereplacepolicy.html) to set on those resources?

https://github.com/awslabs/serverless-application-model/blob/a9388b64147e12426ba585482ee96f81db2dd25a/samtranslator/model/sam_resources.py#L415
https://github.com/awslabs/serverless-application-model/blob/a9388b64147e12426ba585482ee96f81db2dd25a/samtranslator/model/sam_resources.py#L794

**Comment by forzagreen** [comment]
A workaround: 
```bash
cfn-lint --ignore-checks W3011 -t template.yml
```
The error is because SAM creates resources of type `AWS::Lambda::Version` for each lambda function. 
This Version resource has a default property `"DeletionPolicy": "Retain"`.
To see the full resources after transformation run cfn-lint with `--info`.

**Comment by hugoduraes** [comment]
@PatMyron any news on this?

**Comment by ohadbenita** [comment]
Any update on this ?

**Comment by PatMyron** [comment]
[The explicit UpdateReplacePolicy changes](https://github.com/awslabs/serverless-application-model/pull/1481) have been merged into AWS::Serverless transform's develop branch, but [still haven't made it into a release yet](https://github.com/awslabs/serverless-application-model/releases). Will update once those changes have been released

**Comment by PatMyron** [comment]
[Taking longer than expected](https://github.com/awslabs/serverless-application-model/pull/1481#issuecomment-618516935) for [`AWS::Serverless` transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-serverless.html) [`UpdateReplacePolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatereplacepolicy.html) support so [I've exempted those two `AWS::Lambda` resource types ](https://github.com/aws-cloudformation/cfn-python-lint/pull/1529) from being evaluated in [`W3011`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/BothUpdateReplacePolicyDeletionPolicyNeeded.py), so it should now be enforceable as of [version `0.31.1`](https://github.com/aws-cloudformation/cfn-python-lint/pull/1530), which is now available through [`pip3 install cfn-lint --upgrade`](https://pypi.org/project/cfn-lint/0.31.1/)

Apologies for the delays and thanks for the patience

**Comment by mhart** [comment]
I'm still getting this with the below template:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Resources:

  Bucket:
    Type: AWS::S3::Bucket
    DeletionPolicy: Retain
    UpdateReplacePolicy: Retain

  Function:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.handler
      Runtime: nodejs12.x
      Events:
        ObjectCreated:
          Type: S3
          Properties:
            Bucket: !Ref Bucket
            Events: s3:ObjectCreated:*
```

```
[cfn-lint] W3011: Both UpdateReplacePolicy and DeletionPolicy are needed to protect Resources/Bucket from deletion
```
